bbb/services.py

```diff
@@ -379,6 +356,31 @@ def __init__(self, pulse_queue_basename, pulse_exchange_basename, worker_type,
         )
         super(TCListener, self).__init__(*args, events=events, **kwargs)
 
+    def _isAuthorized(self, buildername, scopes):
+        """Tests to see if the builder given is restricted, and if so, whether
+        or not the scopes given are authorized to use it. Builders that do
+        not match the overall restricted builder patterns do not require any
+        scopes. Builders that do must have a buildbot-bridge:builder-name:
+        scope that matches the builder name given."""
+        for r in self.restricted_builders:
+            if re.match(r, buildername):
+                break
+        # If not restricted
+        else:
+            return True
+
+        # If it's a restricted builder, we need to check scopes!
+        for s in scopes:
+            if not s.startswith("buildbot-bridge:builder-name:"):
+                continue
+            allowed_builder = s.split(":")[-1]
+            if buildername == allowed_builder or allowed_builder == "*":
+                return True
+            elif allowed_builder.endswith("*") and buildername.startswith(allowed_builder[:-1]):
+                return True
```

Comment:
Hmm... I think you made that code too complicated... The correct implementation is:
```python
requiredscope = "buildbot-bridge:builder-name:" + buildername
for scope in scopes:
  if scope == requiredscope:
    return True
  if scope.endswith("*") and requiredscope.startswith(scope[:-1]):
    return True
return False
```
But in the noble spirit of making it complicated :)
We could do [like we do in node](https://github.com/taskcluster/taskcluster-base/blob/f1438bee86d28fdf4a66b88e5b32983190e08cbc/utils.js#L50-L68)

```python
requiredscope = "buildbot-bridge:builder-name:" + buildername
return scope_match(scopes, [[requiredscope]])
```
(required scopes for `scope_match` are on disjunctive normal form).
An implementation could look like this:

```python
def scope_match(assumed_scopes, required_scope_sets):
  """
    Take a list of a assumed scopes, and a list of required scope sets on
    disjunctive normal form, and check if any of the required scope sets are
    satisfied.
    
    Example:
    
      required_scope_sets = [
        ["scopeA", "scopeB"],
        ["scopeC"]
      ]
    
    In this case assumed_scopes must contain, either:
    "scopeA" AND "scopeB", OR just "scopeC".
  """
  for scope_set in required_scope_sets:
    for required_scope in scope_set:
      for scope in assumed_scopes:
        if scope == required_scope:
          break   # required_scope satisifed, no need to check more scopes
        if scope.endswith("*") and required_scope.startswith(scope[:-1]):
          break   # required_scope satisifed, no need to check more scopes
      else:
        break     # required_scope not satisfied, don't check reset of scope_set
    else:
      return True # scope_set satisfied, so we're happy
  return False    # none of the required_scope_sets were satisfied
```

I'll file that for inclusion in taskcluster-client.py as it's probably nice to maintained there. For this you can use whatever.